R/ibss.R

```diff
@@ -0,0 +1,182 @@
+    data$XtX %*% colSums(alpha * mu) else NULL
+
+
+  # Return Initialized Model
+  model <- list(alpha        = alpha,
+                mu           = mu,
+                mu2          = mu2,
+                Xr           = Xr,
+                XtXr         = XtXr,
+                KL           = rep(as.numeric(NA), L),
+                lbf          = rep(as.numeric(NA), L),
+                lbf_variable = matrix(as.numeric(NA), L, p),
+                sigma2       = residual_variance,
+                V            = V,
+                pi           = prior_weights
+  )
+
+  class(model) <- "susie"
+
+  # Set null index (for refine step)
+  model$null_index <- if (is.null(null_weight)) 0 else length(prior_weights)
+
+  return(model)
+}
+
+
+### Fitting ###
+
+# This function updates each of the L effects
+
+ibss_fit = function(data, model,
+                    estimate_prior_variance = TRUE,
+                    estimate_prior_method = c("optim"),
+                    check_null_threshold = 0){
+  estimate_prior_method <- match.arg(estimate_prior_method)
+
+  # Repeat for each effect to update
+  L <- nrow(model$alpha)
+  if (L > 0)
+    for (l in seq_len(L)){
+      model <- single_effect_update(data, model, l,
+                                  optimize_V = estimate_prior_method,
+                                  check_null_threshold = check_null_threshold)
+    }
+
+  return(model)
+}
+
+### Finalize ###
+
+# This function takes the final SuSiE model object and appends credible sets,
+# posterior inclusion probabilities, and runs other post-processing steps.
+
+ibss_finalize <- function(data,
+                          model,
+                          coverage               = 0.95,
+                          min_abs_corr           = 0.50,
+                          median_abs_corr        = NULL,
+                          prior_tol              = 1e-9,
+                          n_purity               = 100,
+                          compute_univariate_zscore = FALSE,
+                          intercept              = TRUE,
+                          standardize            = TRUE,
+                          elbo                   = NULL,
+                          iter                   = NA_integer_,
+                          null_weight            = NULL,
+                          track_fit              = FALSE,
+                          tracking               = NULL) {
+
+  # Append ELBO & iteration count to model output
+  model$elbo  <- elbo[2:(iter + 1)]
+  model$niter <- iter
+
+  # Intercept & Fitted Values
+  model$X_column_scale_factors <- get_scale_factors(data)
+  model$intercept              <- get_intercept(data, model, intercept)
+  model$fitted                 <- get_fitted(data, model, intercept)
+
+  # Tracking Across Iterations
+  if (track_fit)
+    model$trace = tracking
+
+  # Credible Sets
+  model$sets <- get_cs(data, model, coverage, min_abs_corr, n_purity) # TODO: Check on median_abs_corr parameter
```

Comment:
We have tried the median_abs_corr as a metric in practice and we did not like it. It seems too rough. Let's stick to just min_abs_corr